_posts/wasmer-go-embedding-1.0.md

```diff
@@ -0,0 +1,543 @@
+slices of values, i.e. it supports multi-values!
+
+```go
+func sum(args []wasmer.Value) ([]wasmer.Value, error) {
+	x := args[0].I32()
+	y := args[1].I32()
+
+	return []wasmer.Value{wasmer.NewI32(x + y)}, nil
+}
+```
+
+Note that a host function manipulates values of kind `Value` to allow
+a more unified API.
+
+We are talking about the host function implementation. To create a
+proper host function, we use the `NewFunction` function. It is also
+possible to create a new host function with
+`NewFunctionWithEnvironment` which expects of a function with the
+following form:
+
+```go
+type MyEnvironment struct {
+	shift int32
+}
+
+environment := &MyEnvironment {
+	shift: 42,
+}
+
+hostFunction := wasmer.NewFunction(
+	store,
+
+	// The host function signature.
+	wasmer.NewFunctionType(
+		wasmer.NewValueTypes(wasmer.I32, wasmer.I32), // two i32 params
+		wasmer.NewValueTypes(wasmer.I32), // one i32 result
+	),
+
+	// Our environment!
+	environment,
+
+	// The function implementation.
+	func(environment interface{}, args []wasmer.Value) ([]wasmer.Value, error) {
+		// Cast to our environment type, and do whatever we want!
+		env := environment.(*MyEnvironment)
+		e := env.shift;
+		x := args[0].I32()
+		y := args[1].I32()
+
+		return []wasmer.Value{wasmer.NewI32(e + x + y)}, nil
+	},
+)
+```
+
+The biggest improvement is that the environment can be anything! No
+more restriction due to `cgo`. The user is free to store anything
+needed.
+
+### A word about exported functions
+
+The `Exports` API provides the following methods: `GetFunction`,
+`GetMemory`, `GetGlobal` and `GetMemory`, to get an external by its
+name (or simply `Get`).
+
+`GetFunction` returns an exported function with a native Go API,
+i.e. a function that can be invoked as `addOne(42)`. It's actually an
+alias of `GetRawFunction(name).Native()` (including error handling):
+
+* A _raw function_ is represented by the type `Function`,
+* A _native function_ is represented by the type `NativeFunction`.
+
+The `Function` type provides more information about the function
+itself, whilst `NativeFunction` serves the purpose of being easy to
+use as any Go functions.
+
+```go
+addOne, _ := instance.Exports.GetRawFunction("add_one")
+
+fmt.Println(addOne.Type())
+fmt.Println(addOne.ParameterArity())
+fmt.Println(addOne.ResultArity())
+
+result, _ := addOne.Call(41)
+// or
+addOneNative := addOne.Native()
+result, _ := addOneNative(41)
+```
+
+This example illustrates that even if the new API is more Go
+idiomatic, it's not at the price of losing information or features.
+
+## Compilers and Engines
+
+Let's talk a moment about compilers and engines. Compilers aim at
+compiling the WebAssembly modules into executable codes. Engines drive
+the compilation _and_ the execution of the WebAssembly modules. This
+design provides a unique flexibility which allows Wasmer to be used in
+various contexts.
+
+### Cranelift has a new companion: Singlepass!
```

Comment:
```suggestion
### Cranelift has new companions: Singlepass and LLVM!
```

I think we should also expose LLVM as it is with a small tldr: "we are about to release Wasmer-LLVM in go soon!"
The main reason for this is that the articles stay "untouched" for long, so we want to present the ideal scenario.

Reply:
I'm not sure LLVM will be embedded inside `libwasmer` _soon_. It has multiple implications. _But_ the Go API already handle LLVM! If we compile `libwasmer` with LLVM, it works. I do it pretty often, and it's tested. I'm updating this section.